### ISSUE TITLE
SQS purge/delete not working

### ISSUE THREAD

**Issue by BeyondEvil** [localstack/localstack#3279]
# Type of request: This is a ...

[x] bug report
[ ] feature request

# Detailed description

I run localstack using the docker container and hit an issue where SQS found "ghost" messages on the queue.

Neither purge-queue or delete-message would remove it, even tho the commands ran successfully.

Deleting the image and associated volume didn't help. What helped was to delete the `recorded_api_calls.json` in the `/tmp/localstack/data` directory.

## Expected behavior

Purge and delete should work as expected.

## Actual behavior

Message can still be received even after it being deleted or entire queue purged.

# Steps to reproduce

## Command used to start LocalStack

docker-compose up

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)

Compose file:

```yml
version: "2.1"
services:

  localstack:
    image: localstack/localstack:0.11.6
    container_name: localstack
    ports:
      - "4566:4566"
    environment:
      - DEBUG=true
      - SERVICES=s3,sns,sqs,dynamodb,secretsmanager
      - DEFAULT_REGION=us-west-2
      - DATA_DIR=/tmp/localstack/data
      - LAMBDA_EXECUTOR=local
      - DOCKER_HOST=unix:///var/run/docker.sock
    volumes:
        - "/tmp/localstack:/tmp/localstack"
    healthcheck:
        test: ["CMD", "curl", "-f", "http://localhost:4566/health"]
        interval: 10s
        timeout: 15s
        retries: 3

  blackbox:
    build:
      context: ./dev
    image: blackbox:latest
    container_name: blackbox
    environment:
      - AWS_DEFAULT_REGION=us-west-2
      - AWS_ACCESS_KEY_ID=test
      - AWS_SECRET_ACCESS_KEY=test
      - CLIENT_SECRET="${CLIENT_SECRET}"
      - LOCALSTACK_ENDPOINT=http://localstack:4566
      - PYTHONDONTWRITEBYTECODE=1
      - CORE_PLATFORM_SNS_TOPIC_ARN=arn:aws:sns:us-west-2:000000000000:CorePlatform
      - DYNAMODB_APP_CREDS_TABLE="stage-sync-app-secret"
    volumes:
      - ${PWD}/dev:/blackbox
    depends_on:
      - localstack

networks:
  default:
    name: blackbox
```

The relevant function that gets the "ghost" message is run inside `blackbox`:

```python
    LOCALSTACK_ENDPOINT = os.getenv("LOCALSTACK_ENDPOINT", "http://localhost:4566")
    boto3.resource = functools.partial(boto3.resource, endpoint_url=LOCALSTACK_ENDPOINT)
    SQS_HEALTH_REPORT_QUEUE_NAME = os.getenv(
        "SQS_HEALTH_REPORT_QUEUE_NAME", "health-report-queue"
    )

    sqs = boto3.resource("sqs", region_name=AWS_DEFAULT_REGION)
    queue = sqs.get_queue_by_name(QueueName=SQS_HEALTH_REPORT_QUEUE_NAME)
    messages = queue.receive_messages(
        MaxNumberOfMessages=1, VisibilityTimeout=60, WaitTimeSeconds=20
    )
    if messages:
        return messages[0]
```

## Environment details:

```
/blackbox # python --version
Python 3.8.6

/blackbox # pip freeze
appdirs==1.4.4
boto3==1.16.18
botocore==1.19.25
certifi==2020.11.8
chardet==3.0.4
distlib==0.3.1
filelock==3.0.12
idna==2.10
jmespath==0.10.0
pipenv==2018.11.26
pycryptodome==3.9.9
python-dateutil==2.8.1
requests==2.25.0
s3transfer==0.3.3
six==1.15.0
urllib3==1.26.2
virtualenv==20.2.1
virtualenv-clone==0.5.4

/blackbox # uname -a
Linux e960a535871b 5.4.39-linuxkit #1 SMP Fri May 8 23:03:06 UTC 2020 x86_64 Linux
```

Host machine: macOS 10.15.7
Docker version 19.03.13, build 4484c46d9d
docker-compose version 1.27.4, build 40524192

**Comment by christopher-nash** [comment]
We experience this as well while running tests on our application. We are using Java and interacting with localstack via the AWS SDK. Before each test is run we issue a purge command against the queue.

Doing some experimenting I narrowed it down to a change between versions `0.11.2` and `0.11.3`:
* `0.11.2` - All tests **pass** reliably
* `0.11.3` - Several tests **fail often**

I also tried it with older versions and in any version _before_ `0.11.3` such as `0.10.7` my tests reliably pass. 
Using other versions _after_ `0.11.3` such as `0.11.7` or `0.12.3` and my tests fail often.

It appears to be some sort of timing/race condition between the purge and the read of messages. I've attempted to recreate the issue using the `aws-cli` manually and haven't been able to, probably because I can't attempt a read quickly enough by hand after the purge.

-----------

Looking at the Localstack release notes for `0.11.3` I see something that may be suspect: 

https://github.com/localstack/localstack/releases/tag/v0.11.3

> 1. New Features
> ...
> * use moto instead of ElasticMQ as default backend for SQS

I don't have a good context for what `moto` is or why it may be causing this issue to periodically happen, but my gut feel is that it was this backend change in `0.11.3` which introduced this issue.

------------

Additional details on our specific project:

AWS SDK version we're using
```xml
<dependency>
  <groupId>software.amazon.awssdk</groupId>
  <artifactId>bom</artifactId>
  <version>2.15.23</version>
  <type>pom</type>
  <scope>import</scope>
</dependency>
```

Running the containers locally via docker run command:
```sh
docker run --name localstack -e SERVICES=sqs -p 4566:4566 -d localstack/localstack:0.11.2
```

The purge command we issue:
```java
private void purgeQueue(String queueUrl) {
    sqsClient.purgeQueue(PurgeQueueRequest.builder()
                           .queueUrl(queueUrl)
                           .build());
}
```

**Comment by irahulranjan** [comment]
Hi @BeyondEvil,

Thanks for reporting the issue.  Does the black box function wait for localstack to restore the infrastructure or starts polling for messages as soon as the health endpoint returns success (services are up and running)?

**Comment by BeyondEvil** [comment]
> Hi @BeyondEvil,
> 
> Thanks for reporting the issue. Does the black box function wait for localstack to restore the infrastructure or starts polling for messages as soon as the health endpoint returns success (services are up and running)?

Honestly, I'm not sure what the state of the logic was at the point where I reported the issue.

Today, it waits for the infrastructure to be restored. But at the time of report, it might've been that it didn't.

Let me see if I can reproduce it with the current state.

**Comment by whummer** [comment]
Thanks for the update @BeyondEvil @christopher-nash . We believe that this should (hopefully) be fixed in #3388 - can you please pull the `latest` Docker image and give it another try? Please report here if the problem persists. Thanks
